scripts/redirectURLParser.js

```diff
@@ -29,10 +31,16 @@ function langURLParser(state) {
 
   // invalid lang but with a valid path.
   if (!langTest && Object.keys(pages).indexOf(pathname) !== -1) {
-    return "/" + validLang + pathname;
+    pathname = "/" + validLang + pathname;
+  } else {
+    pathname = "/" + validLang + "/";
   }
-  // default
-  return "/" + validLang + "/";
+
+  if (redirectQuery) {
+    pathname += query.query || "";
+    query = {};
+  }
+  return {pathname, query};
```

Comment:
This is the bits that used to be in client.jsx `onEnter` now moved in here as explained earlier, easier to test.